proposals/007-azure-kms.md

```diff
@@ -0,0 +1,278 @@
+
+### TLS
+
+With two remote systems involved we want independently customizable TLS for:
+
+1. the authentication client
+2. the key vault client
+
+So that we can supply custom trust or use insecure testing modes (mock servers using self-signed certs for example).
+
+### DEK Generation
+
+Another difference is that basic Key Vault cannot generate the DEK bytes for us. Managed HSM does have an operation to generate random bytes for us https://learn.microsoft.com/en-us/rest/api/keyvault/?view=rest-keyvault-keys-7.4#key-operations-managed-hsm-only
+
+### Key Identifiers
+
+When we wrap/unwrap we have to identify a key by its name and version. You can obtain the latest version by https://learn.microsoft.com/en-us/rest/api/keyvault/keys/get-key/get-key?view=rest-keyvault-keys-7.4&tabs=HTTP.
+
+The API usually returns it as a `kid` https://learn.microsoft.com/en-us/azure/key-vault/general/about-keys-secrets-certificates#object-identifiers. Like
+
+> For Vaults: https://{vault-name}.vault.azure.net/{object-type}/{object-name}/{object-version}
+
+eg. `https://my-vault.vault.azure.net/keys/my-key/78deebed173b48e48f55abf87ed4cf71`
+
+The Azure SDK then takes advantage of the fact this happens to be the base URL for various operations like wrap/unwrap.
+
+ed. for [wrap key](https://learn.microsoft.com/en-us/rest/api/keyvault/keys/wrap-key/wrap-key?view=rest-keyvault-keys-7.4&tabs=HTTP) the url is `POST {vaultBaseUrl}/keys/{key-name}/{key-version}/wrapkey?api-version=7.4
+`
+
+To minimise EDEK bytes there are a couple of opportunities when encoding the EDEK:
+1. we can exclude the vaultBaseUrl and have the limitation that we work only with a single key vault
+2. we can encode just the keyName and keyVersion, and use that to decrypt
+3. all documented examples of the keyVersion (and from my experimentation too) are 128bits encoded as a hexadecimal string.
+
+So for the EDEK I think we could optimistically try to encode it as `versionByte,keyNameLength,keyName,128-bit-decoded,edek`
+and fall back to using the string like `versionByte,keyNameLength,keyName,keyVersionLength,keyVersionString,edek` using
+different version bytes to discriminate between the formats.
+
+## Proposal
+
+1. Support all flavours of Key Vault. The APIs will be the same, just with a different vault base URI.
+2. Support RSA and HSM-RSA key types, wrapping using `RSA-OAEP-256`.
+3. Support HSM-AES key type and AES-GCM wrapping
+4. Support only client credentials oauth flow with Entra using clientId + clientSecret. This supports workloads running anywhere. We could add support for other client credentials (certificates, federated certs) and Managed Identities later. Share a single auth token per Filter Definition until near expiry.
+5. Support TLS customization of the authentication client and key vault client.
+6. If the Key Vault is Managed HSM, then we will use the API to generate random bytes. Else, DEK bytes will be generated proxy-side with a SecureRandom.
```

Comment:
Can we clarify that "the API" means the Key Vault API? (there's multiple APIs at play)
```suggestion
6. If the Key Vault is Managed HSM, then we will use the Key Vault API to generate random bytes. Else, DEK bytes will be generated within the proxy JVM using a `SecureRandom`.
```